### DISCUSSION TITLE
5 giugno 2023 - esercizio 3

### DISCUSSION THREAD

**Discussion by IonutCicio** [sapienzastudentsnetwork/basi-di-dati-1#40] [Esami]
Supponiamo di avere un file di 16.500.000 record. Ogni record occupa 240 byte, di cui 25 per il campo chiave.

Ogni blocco contiene 2048 byte. Un puntatore a blocco occupa 5 byte.

Usiamo una organizzazione B-tree con i **blocchi sia del file principale che del file indice pieni al minimo.**

**Calcolare:**

- il numero di blocchi del file principale

- il numero di blocchi del file indice

- il numero di accessi necessari per ricercare un record del file principale

**Comment by Warcophyr** [comment]
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/b4cf856d-4d0b-48b1-8922-d1badf13c7e6)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/869fab59-8443-4e95-b6c3-747800fce911)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/6df80a5c-2008-4a36-9439-4cf2e9fa7217)
![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/133346488/25d0b7ac-b203-4ac1-82b8-0c7911d84c71)

**Reply by UnoCasuale** [reply]
Procedimento corretto, ma credo tu abbia letto male il key size e abbia preso il record size, quindi poi il risultato non sia corretto

**Comment by MarcoLinardi** [comment]
![IMG_0287](https://github.com/sapienzastudentsnetwork/bdi2324/assets/123003372/4d077108-1113-4fb7-b745-4688250a0f4d)

**Reply by Frisbon** [reply]
Ciao, potresti spiegarmi il perché hai scelto di sottrarre il PS in (a) e aggiungere 1 in (b)?

**Reply by IonutCicio** [reply]
In un B-tree, i blocchi interni sono organizzati nel seguente modo

![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/e83d6c1e-0ab8-4b63-adb1-aa35f0919af8)

dove $p$ sta per "puntatore" e $k$ per chiave

L'idea, quando il riempimento è almeno al 50%, è calcolarsi quante coppie $p$, $k$ _(parte in verde)_ servono per occupare almeno il 50% _(quindi sottraggo l'ultimo puntatore che mi può sballare i conti)_ e poi ce lo aggiungo alla fine.

> [!NOTE]
> Il numero di $p = k + 1$, perché se ho una chiave, devo mettere tutti i record alla sua destra _(maggiori della chiave)_ da qualche parte, quindi mi serve il puntatore dopo

![image](https://github.com/sapienzastudentsnetwork/bdi2324/assets/40034429/396d6147-3811-4f9f-a9bf-fa853fea01c3)
_(rappresentazione un po' più realistica di un blocco pieno almeno al 50%, in genere la chiave ha una dimensione **di gran lunga** maggiore rispetto al puntatore)_

**Comment by Frisbon** [comment]
Non so se é giusto, non capisco bene la teoria perció uso un formulario direttamente.

Esercizio chiede (**pieno al min**) quindi BS = BS **/2** = 1024 e 
RpB = **>**(BS **- PS**/RS) = >(1019/240) = 5

NBpFP = >(NR / RpB) = >(16500000/5) = 33,000,00

m = **<**[(BS-PS)/(KS+PS)] **+1** = <(1019/30) +1 = 35

NBpFI = $\Sigma$ NBpLi

quindi calcolo i livelli, a livello 0 uso NBpFP, ai successivi uso il precedente

NBpL0 = >(NBpFP/m) = 94286
NBpL1 = >(NBpFP0/m) = 2694
77
3
1 ⇒  STOP

Quindi NBpFI = $\Sigma$ NBpLi = 94286 + 2694 +77 + 3 + 1 = 97061

Num Accessi = Altezza albero = 5+1 =6 (Si vede da NBpLi)

**Comment by M1SC0** [comment]
<img width="1356" height="710" alt="image" src="https://github.com/user-attachments/assets/5455a2fc-52a2-486d-a034-8e7dfb513f18" />
